### PR TITLE
feat(material-experimental/mdc-slider): implement MatSlider

### PR DESCRIPTION
**Pull request by wagnermaciel**

### CODE REVIEW

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (crisbeto) [inline src/material-experimental/mdc-slider/slider-thumb-decorator.ts:26]: At least to me, "decorator" implies that it can be mixed together with other decorators, but since this is an element tag, it can't really be mixed. Is there a reason why we don't just call it `mat-slider-start-thumb, mat-slider-end-thumb`?

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (crisbeto) [inline src/material-experimental/mdc-slider/slider-thumb-decorator.ts:31]: Would these need to account for touch devices?

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (crisbeto) [inline src/material-experimental/mdc-slider/slider-thumb-decorator.ts:30]: Who would listen to these events? If they're only here for internal consumption, we could bind them manually from the parent that listens to them.

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (crisbeto) [inline src/material-experimental/mdc-slider/slider-thumb-decorator.ts:45]: Why is this call necessary? Angular should do it by default when an `@Input()` changes, even though it's an `OnPush` component.

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (crisbeto) [inline src/material-experimental/mdc-slider/slider-thumb-decorator.ts:65]: This will hit the DOM unnecessarily on every invocation. The `tagName` of an element can't change so we could resolve it once in the constructor instead.

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:26]: The `MatSliderThumb` Directive (which wraps the sliders' hidden `input` element(s)) already uses the selector `mat-slider-thumb`, which might make it a bit confusing.

We could rename `MatSliderThumb` to `MatSliderInput` and use the selector `mat-slider-input` instead. I believe this was the initial plan, but @jelbourn or @mmalerba wanted it named `mat-slider-thumb` to be more intuitive for developers. That is also when it was decided that the visual slider thumb class would be named `MatSliderThumbDecorator` and have the selector `mat-slider-thumb-decorator`.

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:30]: The plan is to create a child component `MatSliderKnob` which handles all of the ripple states of the MDC slider. It would need to know when the `MatSliderThumbDecorator` is hovered. Is there a different/preferred way to listen for these events from a child component?

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:31]: Probably. Would it be acceptable for me to add a TODO here and implement the support for this in a later PR?

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:65]: This method will be used in the future by the child component `MatSliderKnob` to figure out what thumb it corresponds  to. It will only need to call this method once in its constructor so I don't see it causing any performance issues.

I initially had this value calculated lazily in a getter but I was concerned that I was adding unnecessary complexity to the code for a negligible performance gain.

```ts
get thumb(): Thumb {
    if (!this._thumb) {
        this._thumb = this._getHostElement().tagName === 'MAT-SLIDER-END-THUMB-DECORATOR'
            ? Thumb.END
            : Thumb.START;
    }
    return this._thumb;
}
private _thumb: Thumb;
```

**Comment by wagnermaciel** [comment]
@crisbeto Ready for re-review 👍🏽

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:45]: I'm not sure why it's necessary but I am sure that it's necessary. Without this call, the value indicator text doesn't update properly _while_ dragging - only after the drag ends. Changing the value using arrow keys will not work without this call, either.

Again, I have no idea why this is the case. Let me know if you'd like to take a deeper look and we can schedule a 1:1 to figure out why this happens.

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (mmalerba) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:26]: I didn't so much agree to naming it "decorator" as agreed with the general approach and the naming of the components which constitute the user facing API. I do agree that this name is kind of confusing because it sounds like it might have something to do with the [decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern).

My understanding is that this is a purely internal component, so my first question is, does this component even need to exist at all or can we just roll it into the mat-slider class? If we do think it deserves its own component maybe something like `mat-slider-thumb-ornament`?

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (mmalerba) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:30]: You can always listen to events on an HTML element, regardless of whether its an Angular component or not. My gut feeling is that you can just roll the template for this component into the `<mat-slider>` template and add listeners for elements you care about in `MatSlider`.

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (mmalerba) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:65]: you should call `.toUpperCase()` on `tagName`, I don't think you're guaranteed it will be all caps

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:26]: The main reason I moved it into a separate component was to dedup the html for the start and end thumbs. I can combine it with `MatSlider` by doing something like this instead:
```html
<ng-container *ngIf="isRange">
  <ng-container *ngIf="isDiscrete">
    <div class="mdc-slider__value-indicator-container">
      <div class="mdc-slider__value-indicator">
        <span class="mdc-slider__value-indicator-text">
          {{ startValueIndicatorText }}
        </span>
      </div>
    </div>
  </ng-container>
  <div class="mdc-slider__thumb-knob" #startKnob></div>
</ng-container>

<ng-container *ngIf="isDiscrete">
  <div class="mdc-slider__value-indicator-container">
    <div class="mdc-slider__value-indicator">
      <span class="mdc-slider__value-indicator-text">
        {{ endValueIndicatorText }}
      </span>
    </div>
  </div>
</ng-container>
<div class="mdc-slider__thumb-knob" #endKnob></div>
```

Otherwise, I'm good with renaming it to `MatSliderThumbOrnament` with the selectors `mat-slider-start-thumb-ornament` and `mat-slider-end-thumb-ornament`. Let me know your thoughts

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:30]: I wrote my reason for extracting this into a separate component above. If you decide we'd prefer this to be combined back into the `MatSlider` class, I'll make these changes.

**Comment by wagnermaciel** [comment]
@mmalerba Ready for re-review 👍🏽

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:65]: I'm waiting for your response to whether or not we're going to collapse the `MatSliderThumbDecorator` back into `MatSlider` before I add this change.

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (mmalerba) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:26]: Ah ok, you could probably eliminate the need to copy paste with a `*ngFor="let thumb of thumbs"`. Though I'd also be fine having it extracted into a separate component. I'll leave it up to you, whatever you feel makes it easier.

Comment (jelbourn) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:26]: I would recommend not introducing a separate component since that incurs extra mental overhead vs. just using a for loop

File: src/material-experimental/mdc-slider/slider-thumb-decorator.html
Comment (jelbourn) [inline src/material-experimental/mdc-slider/slider-thumb-decorator.html:5]: Omit spaces in braces in the components codebase

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (jelbourn) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:30]: +1 to doing this in `MatSlider` itself

File: src/material-experimental/mdc-slider/slider-thumb-decorator.html
Comment (crisbeto) [inline src/material-experimental/mdc-slider/slider-thumb-decorator.html:1]: I think you can avoid an extra level of nesting by setting the `*ngIf` on the `mdc-slider__value-indicator-container`.

File: src/material-experimental/mdc-slider/slider-thumb-decorator.ts
Comment (jelbourn) [reply src/material-experimental/mdc-slider/slider-thumb-decorator.ts:45]: +1, it's an anti-pattern to call `detectChanges` like this because it detects changes _immediately_ instead of coallessing multiple changes into a single change detection round

**Review by crisbeto — APPROVED** [review]

**Comment by wagnermaciel** [comment]
@jelbourn @mmalerba Ready for re-review 👍🏽

File: src/material-experimental/mdc-slider/slider.ts
Comment (mmalerba) [inline src/material-experimental/mdc-slider/slider.ts:183]: can we remove this? we don't usually call `markForCheck` in our property setters

File: src/material-experimental/mdc-slider/slider.ts
Comment (mmalerba) [inline src/material-experimental/mdc-slider/slider.ts:207]: Do we need this? why not just have the user get it from the thumb component? (same for `setValue`)

File: src/material-experimental/mdc-slider/slider.ts
Comment (mmalerba) [inline src/material-experimental/mdc-slider/slider.ts:221]: can this be a property instead of a function? (same for the next two)

File: src/material-experimental/mdc-slider/slider.ts
Comment (mmalerba) [inline src/material-experimental/mdc-slider/slider.ts:239]: What are all of these internal methods used for?

Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts:239]: These are just some getters for the `SliderAdapter`

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts:207]: I was also mirroring the old API here. We can also remove this if you'd like

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts:183]: I'm not sure why but without this the component doesn't render correctly. I've already tried removing & redoing this another way before but it still didn't work. Would you be up to do some peer programming to help me debug this?

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts:151]: I was just matching the old API. I can remove them if you'd like

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts:221]: You mean changing `_document` to be a public variable and doing something like this in the constructor?

```ts
this._hostElement = _elementRef.nativeElement;
this._window = _document.defaultView || window;
```

File: src/material-experimental/mdc-slider/slider.ts
Comment (jelbourn) [inline src/material-experimental/mdc-slider/slider.ts:104]:
```suggestion
  get disabled(): boolean { return this._isDisabled; }
```
Just `disabled`, which matches the native API and all the other components (similar for other boolean properties)

File: src/material-experimental/mdc-slider/slider.ts
Comment (jelbourn) [inline src/material-experimental/mdc-slider/slider.ts:121]: What was the rationale for changing this API from the current `MatSlider` (which uses `tickInterval` for this? I also think the API would probably be something like `showTickMarks`

File: src/material-experimental/mdc-slider/slider.ts
Comment (jelbourn) [inline src/material-experimental/mdc-slider/slider.ts:191]: For `document`, follow the pattern used in other components so that the class member is correctly typed to `Document` while still using `any` for the constructor parameter:
```ts
private _document: Document;
constructor(@Inject(DOCUMENT) document: any) {
  this._document = document;
}
```

File: src/material-experimental/mdc-slider/slider.ts
Comment (jelbourn) [reply src/material-experimental/mdc-slider/slider.ts:221]: +1, you should be able to use the `_document` class member

File: src/material-experimental/mdc-slider/slider.ts
Comment (jelbourn) [inline src/material-experimental/mdc-slider/slider.ts:229]: What is this needed for?

File: src/universal-app/kitchen-sink-mdc/kitchen-sink-mdc.html
Comment (jelbourn) [inline src/universal-app/kitchen-sink-mdc/kitchen-sink-mdc.html:110]: In general you should omit any commented code until it's ready to go in for real

File: src/material-experimental/mdc-slider/slider.ts
Comment (mmalerba) [reply src/material-experimental/mdc-slider/slider.ts:151]: It's been a while since we discussed the API. I _think_ we decided that we were going to try to keep most of the API but split it between the container and thumbs. That way people would need to make some changes, but hopefully they would be fairly straightforward, just adding the wrapper and moving a few properties to it. @jelbourn is that what you remember too?

File: src/material-experimental/mdc-slider/slider.ts
Comment (mmalerba) [reply src/material-experimental/mdc-slider/slider.ts:239]: Ah gotcha, I guess the class that uses them is not in this PR

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts:229]: It is used by `SliderAdapter`. It will be changed to a property. See [this comment](https://github.com/angular/components/pull/21680#discussion_r565524803)

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts:57]: I plan to use it for `dragStart`, `dragEnd`, `mouseenter`, and `mouseleave` events. Since it always involves interacting with the slider thumb, we could call it `MatSliderThumbInteractionEvent`

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts:39]: Sgtm. Would you still prefer to move it to its own separate file if we delete the `MatSliderChangeEvent`? See [this comment](https://github.com/angular/components/pull/21680#discussion_r565521866)

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts:207]: Sorry, I was mistaken. Both `getValue` and `setValue` are used by `slider-thumb.ts` in the `value` setter.

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts:121]: I didn't implement `tickInterval` because it is not part of the MDC implementation. We'd have to add that functionality ourselves

I'm ok with renaming this to `showTickMarks` (again, I was just copying the variable name from [the MDC foundation](https://github.com/material-components/material-components-web/blob/master/packages/mdc-slider/foundation.ts#L65))

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts:57]: Side note here, should a value be part of this event? It makes sense for `dragStart` and `dragEnd` but not really for `mouseenter` and `mouseleave`.

File: src/material-experimental/mdc-slider/slider.ts
Comment (mmalerba) [reply src/material-experimental/mdc-slider/slider.ts:194]: That seems odd to me, I would expect the thumbs to be the source of truth about their value - not to have to ask the container

**Comment by wagnermaciel** [comment]
@jelbourn Ready for re-review 👍🏽

File: src/material-experimental/mdc-slider/slider.ts
Comment (jelbourn) [inline src/material-experimental/mdc-slider/slider.ts:39]: I would still move this to its own file like `slider-events.ts`. It would also be good to expand the JsDoc to explain what specifically is covered by "interaction"

File: src/material-experimental/mdc-slider/slider.ts
Comment (jelbourn) [inline src/material-experimental/mdc-slider/slider.ts:84]: Why is the `_initialized` check needed? It looks like `_initialized` is also never being set

File: src/material-experimental/mdc-slider/slider.ts
Comment (jelbourn) [inline src/material-experimental/mdc-slider/slider.ts:138]: I would make this a method rather than a property. In general we only use getters when they're required for either an `@Input` property or for backwards compatibility.

File: src/material-experimental/mdc-slider/slider.html
Comment (jelbourn) [inline src/material-experimental/mdc-slider/slider.html:11]:
```suggestion
    <div *ngFor="let tickMark of tickMarks" [class]="_getTickMarkClass(tickMark)"></div>
```
Prefer property bindings over text interpolation

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts:196]: We could do this for `getValue`, but for `setValue` we need to inform the MDC foundation so it can update the UI

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts:84]: Whoops, this is meant to be set to true `ngAfterViewInit`. I'll fix that.

It's needed because if we want to disable the slider after the component has been initialized we have to also inform the foundation with `this._foundation.setDisabled(v);`. And we can't just call it every time because if the foundation hasn't been initialized it will throw errors

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts]: I made this change but it creates a circular dependency

**Comment by wagnermaciel** [comment]
@jelbourn Ready for re-review 👍🏽

File: src/material-experimental/mdc-slider/slider.ts
Comment (jelbourn) [reply src/material-experimental/mdc-slider/slider.ts:76]: Ack, can you add a comment explaining that?

Also, adding unit tests as you go is the thing that typically catches oversights like this. I'd recommend doing future additions alongside with unit tests, otherwise they'll start to get hard to manage.

File: src/material-experimental/mdc-slider/slider.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider.ts]: For the record, we looked into this together and found that we will need to trigger change detection manually when `tickMarks` is updated, but we will instead be triggering change detection from the adapter rather than in the setter. In preparation for that, I've made `_cdr` a public variable so that we can call it from `updateTickMarks` in `slider-adapter.ts` once it's implemented

**Comment by wagnermaciel** [comment]
@jelbourn Ready for re-review 👍🏽

**Review by jelbourn — APPROVED** [review]
LGTM

File: src/material-experimental/mdc-slider/slider.ts
Comment (jelbourn) [reply src/material-experimental/mdc-slider/slider.ts]: Let's go with that since there are relatively few usages in Google so it should be pretty straightforward to swap

File: src/material-experimental/mdc-slider/slider.ts
Comment (mmalerba) [inline src/material-experimental/mdc-slider/slider.ts:139]: These should probably go on the individual thumbs right?

File: src/material-experimental/mdc-slider/slider-thumb.ts
Comment (mmalerba) [inline src/material-experimental/mdc-slider/slider-thumb.ts:65]: Does this need to be its own property? you can just use `this._elementRef.nativeElement`

File: src/material-experimental/mdc-slider/slider.ts
Comment (mmalerba) [inline src/material-experimental/mdc-slider/slider.ts:177]: `@Inject(DOCUMENT) doc: any` (no need to save it as a property since you're assigning it to `this._document`)

File: src/material-experimental/mdc-slider/slider.ts
Comment (mmalerba) [reply src/material-experimental/mdc-slider/slider.ts:198]: can you at least add an `_` to mark these as internal

File: src/material-experimental/mdc-slider/slider.ts
Comment (mmalerba) [reply src/material-experimental/mdc-slider/slider.ts:217]: I think maybe I'll have some organization comments when you send out the PR with the adapter implementation, but for now this is fine

File: src/material-experimental/mdc-slider/slider-thumb.ts
Comment (wagnermaciel) [reply src/material-experimental/mdc-slider/slider-thumb.ts:65]: It doesn't need to. I can also delete the `_hostElement` convenience variable in `slider-thumb.ts`

**Review by mmalerba — APPROVED** [review]
LGTM

**Pull request merged**
